soc: nrf53: allow configuring oscillators from DT

gmarull opened the pull request:
Check commit messages.

gmarull committed:
dts: bindings: clock: add nordic,nrf53-lfxo

Add binding for the nRF53 series LF crystal oscillator.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: clock: add nordic,nrf53-hfxo

Add binding for the nRF53 series HF crystal oscillator.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: nordic: nrf5340: adjust OSCILLATORS IP description

- Create a new compatible: nordic,nrf53x-oscillators, as other series,
  e.g. nRF54LX contain a similar but different IP (with PLL control,
  etc.)
- Adjust DT: use recommended node name, remove redundant status okay.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: nordic: nrf5340: instantiate LF crystal oscillator

LFXO is represented as a child of the oscillators node. A new node is
created because it requires its own properties (see the binding for more
details).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: nordic: nrf5340: instantiate HF crystal oscillator

HFXO is represented as a child of the oscillators node. A new node is
created because it requires its own properties (see the binding for more
details).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf53: allow configuring L|HFXO from DT

Support both, Kconfig (about to be deprecated) and DT.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: bl5340_dvk: remove redundant LFXO option

7pF is already the default.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf53*: add L|HFXO configurations

- Only one board tweaked the HFXO configuration (bl5340_dvk)
- LFXO configuration has been added to all boards based on the current
  Kconfig defaults: use LFXO with internal 7pF capacitor.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: nordic: nrf53: deprecate all L|HFXO options

Devicetree should be used instead. Example DT snippets are provided to
ease with the transition.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

carlescufi reviewed (changes requested):
Needs an entry in the migration guide

masz-nordic commented on dts/bindings/clock/nordic,nrf53x-hfxo.yaml:
Is separate binding (instead of using `nordic,nrf-hfxo.yaml`) needed in order to limit enum values?

gmarull replied:
yes, Zephyr bindings do not have advances capacilities like e.g. dt-schema where a single binding could be used

masz-nordic commented on dts/bindings/clock/nordic,nrf53x-oscillators.yaml:
I think this is not the right path - with this approach we'd need to start introducing `nordic,nrf53x-uarte`, etc.
Presence of HW features could be described in properties of generic node instead?

gmarull replied:
This follows more closely the DT-spec, where compatible implies one or another programming model. Zephyr is not consistent on this, though. Some vendors use the `v1`, `v2`, etc. scheme (ST/GD IIRC).

gmarull commented on dts/bindings/clock/nordic,nrf53x-lfxo.yaml:
```diff
@@ -0,0 +1,30 @@
+  clock-frequency:
+    const: 32768
+
+  load-capacitors:
+    type: string
+    enum:
+      - "internal"
+      - "external"
+    description: |
+      Type of load capacitors connected to the crystal.
+
+  load-capacitance-picofarad:
+    type: int
+    enum:
+      - 6
+      - 7
+      - 9
+    description: |
+      Load capacitance in picofarads. This property is only used when
+      load-capacitors is set to "internal".
```
we have 3 possible states:

1. no load-capacitors specified: do nothing and rely on retained content
2. internal: use load-capacitance-picofarad
3. external: use external caps

with load-capacitance-picofarad alone, you can't determine the (1) case (present would be use internal, missing external, but 1st case?)

bjarki-andreasen replied:
I don't think the first state should be allowed :) Either you use internal caps, and if you do, you must specify their value, or you do not, at which case they are disconnected. Is sounds like option 1 is to simply not initialize the clock, that should not be allowed right?

gmarull replied:
these are retained registers according to the spec.

bjarki-andreasen replied:
ok, so the idea is that a bootloader of TF-M will set these values, and zephyr will not update the value on init if` load-capacitors` is not specified. Am I wrong, and is this documented somewhere if not? :)

gmarull replied:
I'm not sure about this use case TBH, I just tried to preserve the functionality :-)

gmarull committed:
doc: releases: migration-guide: add nRF53 L|HFXO changes

Infom users about Kconfig options deprecation, and provide code snippets
on what they have to do now.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> Needs an entry in the migration guide

@carlescufi done

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -490,7 +517,9 @@ static int nordicsemi_nrf53_init(void)
 #endif /* defined(CONFIG_SOC_ENABLE_LFXO) */
-#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL)
+#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL) || \
+	(defined(NRF_SOC_SECURE_SUPPORTED) && \
```
```suggestion
	(defined(CONFIG_NRF_SOC_SECURE_SUPPORTED) && \
```

gmarull replied:
fixed

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -490,7 +517,9 @@ static int nordicsemi_nrf53_init(void)
-#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL)
+#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL) || \
+	(defined(NRF_SOC_SECURE_SUPPORTED) && \
+	 DT_ENUM_HAS_VALUE(DT_NODELABEL(hfxo), load_capacitors, internal))
```
```suggestion
	 DT_ENUM_HAS_VALUE(HXFO_NODE, load_capacitors, internal))
```

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -515,13 +544,15 @@ static int nordicsemi_nrf53_init(void)
 	nrf_oscillators_hfxo_cap_set(NRF_OSCILLATORS, true, capvalue);
-#elif defined(CONFIG_SOC_HFXO_CAP_EXTERNAL)
+#elif defined(CONFIG_SOC_HFXO_CAP_EXTERNAL) || \
+	DT_ENUM_HAS_VALUE(DT_NODELABEL(hfxo), load_capacitors, external)
```
```suggestion
	DT_ENUM_HAS_VALUE(HXFO_NODE, load_capacitors, external)
```

gmarull replied:
changed

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -472,15 +503,11 @@ static int nordicsemi_nrf53_init(void)
-			NRF_OSCILLATORS_LFXO_CAP_9PF :
-			NRF_OSCILLATORS_LFXO_CAP_EXTERNAL);
+#ifdef CONFIG_SOC_NRF5340_CPUAPP
+#if defined(CONFIG_SOC_ENABLE_LFXO) || \
+	DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, external) || \
+	DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, internal)
```
Why not just `#if defined(LFXO_CAP)`?

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -490,7 +517,9 @@ static int nordicsemi_nrf53_init(void)
 #endif /* !defined(CONFIG_BUILD_WITH_TFM) */
 #endif /* defined(CONFIG_SOC_ENABLE_LFXO) */
-#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL)
+#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL) || \
+	(defined(NRF_SOC_SECURE_SUPPORTED) && \
+	 DT_ENUM_HAS_VALUE(DT_NODELABEL(hfxo), load_capacitors, internal))
```
Also here `#if defined(HFXO_CAP_VAL_X2)` could be used (`HFXO_CAP_VAL_X2` could be defined only when `CONFIG_NRF_SOC_SECURE_SUPPORTED`).

gmarull replied:
thanks for the optimization, updated

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
`&hfxo`

gmarull replied:
fixed

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
`&hfxo`

gmarull replied:
fixed

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -258,6 +310,13 @@ config SOC_HFXO_CAP_INT_VALUE_X2
 	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
`&hfxo`

gmarull replied:
fixed

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
 	depends on NRF_SOC_SECURE_SUPPORTED
+	select DEPRECATED
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
+	      load-capacitors = "internal";
+	      ...
+	  };
```
Maybe it would be better to refer to the SOC_HFXO_CAP_INT_VALUE_X2 option regarding the example configuration instead?

gmarull replied:
added a comment

fabiobaltieri commented:
@gmarull needs a rebase as well

gmarull commented:
> @gmarull needs a rebase as well

done

anangl commented on soc/nordic/nrf53/Kconfig:
```diff
@@ -215,19 +219,49 @@ config SOC_ENABLE_LFXO
 choice SOC_LFXO_LOAD_CAPACITANCE
 	prompt "LFXO load capacitance"
 	depends on SOC_ENABLE_LFXO
-	default SOC_LFXO_CAP_INT_7PF
```
Why is this removed? Behavior of these deprecated options should not be changed.

gmarull replied:
likely a mistake, re-added

anangl commented on dts/bindings/clock/nordic,nrf53x-lfxo.yaml:
```diff
@@ -0,0 +1,30 @@
+
+description: Nordic nRF low-frequency crystal oscillator (nRF53X series)
+
+compatible: "nordic,nrf53x-lfxo"
```
It's nRF53 Series, not nRF53X, so I think the compatibles should be "nordic,nrf53-*". Actually, we already have "nordic,nrf53-flash-controller" in tree.

gmarull replied:
we actually have `SOC_SERIES_NRF53X`, also to be consistent with 52 where we have 52x.

gmarull replied:
So please let me know what you prefer, I have no strong opinion here.

nordicjm replied:
For dts I'm not bothered either way, the problem is with the Kconfig symbol. Can see here the series is nrf53 https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/soc.yml#L16 which means https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/nrf53/Kconfig.soc#L6 should be SOC_SERIES_NRF53 not SOC_SERIES_NRF53X, this will completely break when the symbols are automatically generated by the build system, it hasn't been changed yet because that symbol is used in a lot of places, including the HAL, so that will need a big update. I would lean on nrf53 for dts, since nrf53x is an issue with the Kconfig symbol

gmarull replied:
updated to `nrf53`

anangl commented on dts/bindings/clock/nordic,nrf53x-oscillators.yaml:
Name of this file should be corrected.

gmarull replied:
oops, done

nika-nordic commented on dts/bindings/clock/nordic,nrf53-hfxo.yaml:
```diff
@@ -0,0 +1,54 @@
+include: fixed-clock.yaml
+
+properties:
+  clock-frequency:
```
I think it would be beneficial to introduce common compatible like `nordic,nrf53-clock.yaml`, that would contain common `clock-frequency` and `load-capacitors` properties, to be utilized by both `nrf53-hfxo` & `nrf53-lfxo`.
`clock-frequency` could be extracted to even more generic `nordic-nrf5x-clock.yaml` . This way we would increase number of bindings, but reduce duplication across bindings

gmarull replied:
`clock-frequency` is already common (comes from `fixed-clock`), so the only one that can be made common is `load-capacitors`.
